### PR TITLE
WIP PERF: better cache macro expansions

### PR DESCRIPTION
**Pull request by vlad20012**

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/utils/ModificationTrackers.kt [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+private class AndModificationTracker(
+    private val left: ModificationTracker,
+    private val right: ModificationTracker
+) : ModificationTracker {
+    private var lastLeftModCount: Long = -1
+    private var lastRightModCount: Long = -1
+
+    override fun getModificationCount(): Long {
+        val newLeftModCount = left.modificationCount
+        if (newLeftModCount == lastLeftModCount) return lastRightModCount
```
Comment (ortem) [inline src/main/kotlin/org/rust/lang/utils/ModificationTrackers.kt]: Did you really mean `return lastRightModCount`, not `return lastLeftModCount`?

File: src/main/kotlin/org/rust/lang/utils/ModificationTrackers.kt [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        lastLeftModCount = newLeftModCount
+
```
Comment (ortem) [inline src/main/kotlin/org/rust/lang/utils/ModificationTrackers.kt]: Maybe replace
```
val newRightModCount = right.modificationCount
if (newRightModCount == lastRightModCount) return lastRightModCount

lastRightModCount = newRightModCount
return newRightModCount
```
with
```
val newRightModCount = right.modificationCount
lastRightModCount = newRightModCount
return newRightModCount
```
or just
```
lastRightModCount = right.modificationCount
return lastRightModCount
```

File: src/main/kotlin/org/rust/lang/utils/ModificationTrackers.kt (hunk 1 above)
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/utils/ModificationTrackers.kt]: Yep ;)

File: src/main/kotlin/org/rust/lang/core/macros/MacroExpansion.kt [hunk 3]
```diff
@@ -38,7 +38,30 @@ fun expandMacro(call: RsMacroCall): CachedValueProvider.Result<List<ExpansionRes
...
-            CachedValueProvider.Result.create(result, project.rustStructureModificationTracker)
+
+            // The cached value will be invalidated in these cases:
+            // 1. Some descendant of RsMacroCall PSI is changed
+            // 2. Some descendant of RsMacro PSI is changed
+            // 3. RsMacroCall starts to refer to another RsMacro definition
+            //
+            // The interesting case is when macro call or definition is stubbed.
+            // In this case PSI modification tracking is not provided. But subbed
+            // PSI updates not incrementally. I.e. if something changed in a stubbed
```
Comment (fedochet) [inline src/main/kotlin/org/rust/lang/core/macros/MacroExpansion.kt]: `is not updated incrementally` maybe?

File: src/main/kotlin/org/rust/lang/core/macros/MacroExpansion.kt [hunk 4]
```diff
@@ -38,7 +38,30 @@ fun expandMacro(call: RsMacroCall): CachedValueProvider.Result<List<ExpansionRes
...
+            // file, all its PSI will be completely invalidated and new one built.
+            // So there is a 2 cases:
```
Comment (fedochet) [inline src/main/kotlin/org/rust/lang/core/macros/MacroExpansion.kt]: `there are 2 cases`

Comment (fedochet) [reply src/main/kotlin/org/rust/lang/core/macros/MacroExpansion.kt]: It is better like this (IMO):
```
So there are two possible scenarios:
1. Changed file is a file with a macro call. In this case, the cache attached ...
2. Changed file is a stubbed file containing a macro definition referred by some macro call. In this case, ...
```

File: src/main/kotlin/org/rust/lang/utils/ModificationTrackers.kt [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+package org.rust.lang.utils
+
+import com.intellij.openapi.util.ModificationTracker
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiReference
+import org.rust.lang.core.psi.rustStructureModificationTracker
+import java.lang.ref.WeakReference
+
+/**
+ * This modification tracker hold 2 values: the psi reference and it's target.
```
Comment (fedochet) [inline src/main/kotlin/org/rust/lang/utils/ModificationTrackers.kt]: its*

File: src/main/kotlin/org/rust/lang/utils/ModificationTrackers.kt [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            return ++modCount
+        }
+
+        return modCount
+    }
+
+    companion object {
+        /**
+         * Creates [ReferenceTargetModificationTracker] for some reference that
+         * dependent on [rustStructureModificationTracker] (i.e. some reference,
```
Comment (fedochet) [inline src/main/kotlin/org/rust/lang/utils/ModificationTrackers.kt]: `depends` maybe?

File: src/main/kotlin/org/rust/lang/utils/ModificationTrackers.kt [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        }
+    }
+}
+
+/**
+ * This modification tracker increments only if both trackers are incremented.
+ * The main feature of this modification tracker is that it is lazy. I.e. it will
+ * not check modification count of the second tracker if the count of the first tracker
+ * is remained the same. It may be useful if computation of the modification count
+ * of the second tracked is too expensive.
```
Comment (fedochet) [inline src/main/kotlin/org/rust/lang/utils/ModificationTrackers.kt]: `tracker`, not `tracked`

File: src/main/kotlin/org/rust/lang/core/macros/MacroExpansion.kt [hunk 8]
```diff
@@ -38,7 +38,30 @@ fun expandMacro(call: RsMacroCall): CachedValueProvider.Result<List<ExpansionRes
...
-            CachedValueProvider.Result.create(result, project.rustStructureModificationTracker)
+
+            // The cached value will be invalidated in these cases:
+            // 1. Some descendant of RsMacroCall PSI is changed
+            // 2. Some descendant of RsMacro PSI is changed
+            // 3. RsMacroCall starts to refer to another RsMacro definition
+            //
+            // The interesting case is when macro call or definition is stubbed.
+            // In this case PSI modification tracking is not provided. But stubbed
+            // PSI is not updated incrementally. I.e. if something changed in a stubbed
```
Comment (ortem) [inline src/main/kotlin/org/rust/lang/core/macros/MacroExpansion.kt]: "if something changed" -> "if something is changed"

**Comment by vlad20012** [comment]
More tests needed. Except this, everything is ready for review

**Comment by vlad20012** [comment]
Will be implemented in another way

**Pull request closed without merging**